### PR TITLE
[new+change+fix] Expanding Array creation Module

### PR DESCRIPTION
**Pull request by shivasankarka**
## New

- To implement all the array creation routines from numpy and expand the NuMojo array creation functionality (Most have been implemented now, few more left to implement).

## Changes

- To make matmul flexible for different shapes of input arrays. Since NuMojo uses a convention of ignoring the single index in shape i.e `(3, 2, 1)` becomes `(3, 2)`, The matmul needs to be modified to account for this in cases like 
```
A(shape = MxN) X B(shape = N)
```
- Change the name of the buffer of the `NDArray` to `_buf`.
- Change the way to get the shape of the array. After change, `array.shape` returns the shape of array as `NDArrayShape`.

## Fix

- Remove `inout` before self for `NDArray.__getitem__`.

### CODE REVIEW

**Comment by forFudan** [comment]
>Since NuMojo uses a convention of ignoring the single index in shape i.e (3, 2, 1) becomes (3, 2)

Do we have this feature already in the code? I believe that `nm.zeros(Shape(3,2,1))` still generate a array with shape 3x2x1 instead of 3x2.

**Comment by shivasankarka** [comment]
> Do we have this feature already in the code? I believe that `nm.zeros(Shape(3,2,1))` still generate a array with shape 3x2x1 instead of 3x2.

Let's say we return a slice (`arr[:,:,0:1]`) of a 3x4x5 array, the resulting array is returned as `3x4` instead of `3x4x1`. Most of the function in NuMojo follow this convention right now.

**Comment by forFudan** [comment]
> Let's say we return a slice (`arr[:,:,0:1]`) of a 3x4x5 array, the resulting array is returned as `3x4` instead of `3x4x1`. Most of the function in NuMojo follow this convention right now.

This is interesting. `0:1` is somehow interpreted as an integer.

**Comment by forFudan** [comment]
> @forFudan , Yes, the slices with those consecutive integers are interpreted that way. Now I am wondering which convention would be best to follow. In the current convention, there's no difference between a `Nx1` and `1xN` matrix (They are interpreted as a size N vector).

Each Int decrease the dimension by 1. So slide 0:1 should not. This is the convention of numpy and I found this behavior well defined.

File: numojo/core/array_creation_routines.mojo [hunk 1]
```diff
@@ -790,6 +790,40 @@ fn triu[dtype: DType = DType.float64](m: NDArray[dtype], k: Int = 0) raises -> N
     return result^
 
 
+fn vander[dtype: DType = DType.float64](x: NDArray[dtype], N: Optional[Int] = None, increasing: Bool = False) raises -> NDArray[dtype]:
```
Comment (forFudan) [inline numojo/core/array_creation_routines.mojo]: For `N`, maybe we can use the `N: Int = -1` trick 😃

`fn vander[dtype: DType = DType.float64](x: NDArray[dtype], N: Int = -1, increasing: Bool = False)`

If it is -1, then return a square array.

In this case we do not need to use `Option`.

File: tests/test_array_creation.mojo [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import numojo as nm
 from numojo.prelude import *
+from numojo.prelude import *
```
Comment (forFudan) [inline tests/test_array_creation.mojo:3]: This is duplicated.

**Comment by forFudan** [comment]
@shivasankarka I have an interesting suggestion. Shall we just rename `NDArrayShape` to `Shape`, since there is no conflicting definition? Also, rename `NDArray.ndshape` as `NDArray.shape`.

**Comment by MadAlex1997** [comment]
> @shivasankarka I have an interesting suggestion. Shall we just rename `NDArrayShape` to `Shape`, since there is no conflicting definition? Also, rename `NDArray.ndshape` as `NDArray.shape`.

[ndarray.shape in numpy](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.shape.html) is a tuple representing the shape. Having an iterable thing bound to either `ndarray.shape` or `ndarray.shape()` is more convenient than having to go `ndarray.shape.buffer`.

**Comment by forFudan** [comment]
> Having an iterable thing bound to either ndarray.shape or ndarray.shape() is more convenient than having to go ndarray.shape.buffer.

Currently, `NDArrayShape` is already subscripable, which means we can already do this now:

```mojo
A.ndshape[0]
```

So my proposal is to make it something like:

```mojo
A.shape[0]
```

**Comment by shivasankarka** [comment]
> @shivasankarka I have an interesting suggestion. Shall we just rename `NDArrayShape` to `Shape`, since there is no conflicting definition? Also, rename `NDArray.ndshape` as `NDArray.shape`.

I don't see a problem with that change right now. To my knowledge, numpy exposes it as `np.ndarray.shape()`, so that change should be fine. The only thing that might be confusing (for internal developing and debugging purposes) is having a syntax like `arr(shape=Shape(...))` instead of `arr(shape=NDArrayShape(...)). The argument `shape` will keep repeating a lot.

**Comment by forFudan** [comment]
>To my knowledge, numpy exposes it as np.ndarray.shape()
@shivasankarka `numpy` did it via `np.ndarray.shape` member (not method). We currently do this with `NDArray.ndshape` and `NDArray.shape()`.

**Review by forFudan — APPROVED** [review]

**Pull request merged**